### PR TITLE
Rstudio integrate pool

### PR DESCRIPTION
**Pull request by ramirobg94**

### CODE REVIEW

**Review by SteRiccio — APPROVED** [review]

**Review by minotogna — COMMENTED** [review]
does in localhost use local rStudio?

**Comment by ramirobg94** [comment]
> does in localhost use local rStudio?

if we have RSTUDIO_SERVER_URL we use this if not the pool service!

File: webapp/views/App/views/Analysis/Chain/store/actions/chain/useOpenRStudio.js [hunk 1]
```diff
@@ -5,15 +5,22 @@ import { useDispatch } from 'react-redux'
...
+  const { data = {} } = await axios.post('/api/rstudio')
+  const { instanceId = false } = data
+
+  if (instanceId && ProcessUtils.ENV.rStudioProxyServerURL) {
+    return `${ProcessUtils.ENV.rStudioProxyServerURL}${instanceId}_${userUuid}`
   }
+
   if (ProcessUtils.isEnvDevelopment) {
     return 'http://localhost:8787'
   }
```
Comment (minotogna) [inline webapp/views/App/views/Analysis/Chain/store/actions/chain/useOpenRStudio.js:26]: this should not be hardcoded. take it from env variables

**Review by minotogna — CHANGES_REQUESTED** [review]

File: .env.template [hunk 2]
```diff
@@ -30,7 +30,11 @@ SESSION_ID_COOKIE_SECRET=my-cookie-secret-key
 USE_HTTPS=false
 
 # RStudio Server
-RSTUDIO_SERVER_URL=http://localhost:8787
+RSTUDIO_DOWNLOAD_SERVER_URL=
```
Comment (minotogna) [inline .env.template:33]: since this is our domain, any other way we can get it dynamically ?

File: server/modules/analysis/api/chainApi.js [hunk 3]
```diff
@@ -162,12 +165,25 @@ export const init = (app) => {
...
+        await FileUtils.copyDir({ source: rChain._dir, destination: `/tmp/${surveyId}_${chainUuid}_${folderToken}/` })
+        res.json({ folderToken, serverUrl })
       } catch (error) {
         next(error)
       }
     }
   )
+
+  // === Download R SCRIPTS
+  app.get('/download/survey/:surveyId/processing-chain/:chainUuid/script', (req, res, next) => {
```
Comment (minotogna) [inline server/modules/analysis/api/chainApi.js:179]: for convention I would set his as `/survey/:surveyId/processing-chain/:chainUuid/scrip/download`

File: server/modules/analysis/api/chainApi.js [hunk 4]
```diff
@@ -162,12 +165,25 @@ export const init = (app) => {
...
+    try {
```
Comment (minotogna) [inline server/modules/analysis/api/chainApi.js:180]: also, dont we need to generate a temp uuid to validate it?

**Review by minotogna — CHANGES_REQUESTED** [review]

**Comment by minotogna** [comment]
> > does in localhost use local rStudio?
> 
> if we have RSTUDIO_SERVER_URL we use this if not the pool service!

nice! thanks

File: server/modules/analysis/api/chainApi.js (hunk 4 above)
Comment (ramirobg94) [reply server/modules/analysis/api/chainApi.js:180]: The scripts are moved to a folder with a UUID which is the "Token" and after download it, we deleted the folder.

File: .env.template (hunk 2 above)
Comment (ramirobg94) [reply .env.template:33]: this was to put the tunnel to localhost to download from RStudio, in production shouldn't be a problem.
now this URL is taken automatically with the request -> const serverUrl = Request.getServerUrl(req), but in local I had to put [ngrok](https://ngrok.com/) to be able to download the Code.

I can remove this config if you consider that it is cleaner

File: server/modules/analysis/api/chainApi.js (hunk 3 above)
Comment (ramirobg94) [reply server/modules/analysis/api/chainApi.js]: Done!

Comment (ramirobg94) [reply server/modules/analysis/api/chainApi.js]: I did public all of the routes with download, should we add some subfix or something?

Comment (minotogna) [reply server/modules/analysis/api/chainApi.js]: /public ?

File: .env.template (hunk 2 above)
Comment (minotogna) [reply .env.template:33]: add a comment, and add it to env.template

File: webapp/views/App/views/Analysis/Chain/store/actions/chain/useOpenRStudio.js [hunk 5]
```diff
@@ -5,35 +5,88 @@ import { useDispatch } from 'react-redux'
...
+  const rStudioPoolUrl = await _getRStudioPoolUrl({ userUuid })
+
+  if (rStudioPoolUrl) {
+    return rStudioPoolUrl
   }
+
   return `${window.location.origin}/rstudio/`
 }
 
+const _getTStudioCode = ({ surveyId, chainUuid, folderToken, serverUrl }) =>
```
Comment (minotogna) [inline webapp/views/App/views/Analysis/Chain/store/actions/chain/useOpenRStudio.js:43]: what is TStudio?

File: webapp/views/App/views/Analysis/Chain/store/actions/chain/useOpenRStudio.js [hunk 6]
```diff
@@ -5,35 +5,88 @@ import { useDispatch } from 'react-redux'
...
+    ProcessUtils.ENV.rStudioDownloadServerUrl || serverUrl
+  }/api/survey/${surveyId}/processing-chain/${chainUuid}/script/download?surveyCycleKey=0&folderToken=${folderToken}';download.file(url,"./${folderToken}.zip");unzip("./${folderToken}.zip",exdir=".");file.remove("./${folderToken}.zip")`
+
+const _copyRStudioCode = ({ rStudioCode }) => {
+  const input = document.body.appendChild(document.createElement('input'))
+  input.value = rStudioCode
+  input.focus()
+  input.select()
+  document.execCommand('copy')
+  input.remove()
```
Comment (minotogna) [inline webapp/views/App/views/Analysis/Chain/store/actions/chain/useOpenRStudio.js:54]: can we add a domUtils copyToClipboard ?

Comment (ramirobg94) [reply webapp/views/App/views/Analysis/Chain/store/actions/chain/useOpenRStudio.js]: done

File: webapp/views/App/views/Analysis/Chain/store/actions/chain/useOpenRStudio.js (hunk 5 above)
Comment (ramirobg94) [reply webapp/views/App/views/Analysis/Chain/store/actions/chain/useOpenRStudio.js]: typo, fixed!

File: server/modules/analysis/api/chainApi.js (hunk 3 above)
Comment (ramirobg94) [reply server/modules/analysis/api/chainApi.js]: done

File: .env.template (hunk 2 above)
Comment (ramirobg94) [reply .env.template:33]: done

File: server/utils/response.js [hunk 7]
```diff
@@ -49,3 +47,25 @@ export const setContentTypeFile = (res, fileName, fileSize = null, contentType =
...
+}
+
+export const sendZipFile = (res, dir, name) => {
+  if (FileUtils.existsDir(dir)) {
+    setContentTypeFile(res, name, null, contentTypes.zip)
+
+    const zip = Archiver('zip')
+    zip.pipe(res)
+    zip.directory(dir, false)
+    zip.finalize()
```
Comment (SteRiccio) [inline server/utils/response.js:64]: it can be just `Archiver('zip').pipe(res).directory(dir, false).finalize()`

**Review by SteRiccio — DISMISSED** [review]

**Review by SteRiccio — DISMISSED** [review]

File: server/utils/response.js (hunk 7 above)
Comment (ramirobg94) [reply server/utils/response.js:64]: default(...)(...).pipe(...).directory is not a function

**Pull request merged**